// tslint-to-eslint-config/src/converters/lintConfigs/rules/ruleMergers/no-use-before-define.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeNoUseBeforeDefine = void 0;
const lodash_1 = require("lodash");
const NO_USE_BEFORE_DEFINE_DEFAULT_OPTS = {
    classes: true,
    enums: true,
    functions: true,
    ignoreTypeReferences: true,
    typedefs: true,
    variables: true,
};
const mergeNoUseBeforeDefine = (existingOptions, newOptions) => {
    const existingSingleOption = existingOptions === null || existingOptions === void 0 ? void 0 : existingOptions[0];
    const newSingleOption = newOptions === null || newOptions === void 0 ? void 0 : newOptions[0];
    if (existingSingleOption === undefined && newSingleOption === undefined) {
        return [];
    }
    // when not explicitly opting out with a flag assume that the option is requesting at least that flag
    if ((existingSingleOption === undefined && isOnlyTrueFlags(newSingleOption)) ||
        (newSingleOption === undefined && isOnlyTrueFlags(existingSingleOption))) {
        return [];
    }
    return [
        lodash_1.assignWith({}, NO_USE_BEFORE_DEFINE_DEFAULT_OPTS, existingSingleOption, newSingleOption, (currentValue, newValue) => {
            if (currentValue === undefined) {
                return newValue;
            }
            // when merging the flags always keep the opted out (false) flags
            return !!currentValue && !!newValue;
        }),
    ];
};
exports.mergeNoUseBeforeDefine = mergeNoUseBeforeDefine;
function isOnlyTrueFlags(flagsObject) {
    return lodash_1.isObject(flagsObject) && Object.values(flagsObject).every((flag) => !!flag);
}
